src/main/java/picard/sam/AbstractAlignmentMerger.java

```diff
@@ -751,54 +765,104 @@ private void transferAlignmentInfoToPairedRead(final SAMRecord firstUnaligned, f
+                    negClipFrom = negClipFrom > 0 ? (neg.getReadLength() + 1) - negClipFrom : 0;
 
-                    if (negDiff > 0) {
-                        final int clipped = lengthOfSoftClipping(neg.getCigar().getCigarElements().iterator());
-                        final int clipFrom = neg.getReadLength() - negDiff - clipped + 1;
-                        CigarUtil.softClip3PrimeEndOfRead(neg, Math.min(neg.getReadLength(), clipFrom));
+                    if(posClipFrom > 0) {
+                        clip3PrimeEndOfRead(pos, posClipFrom, useHardClipping);
+                    }
+                    if(negClipFrom > 0) {
+                        clip3PrimeEndOfRead(neg, negClipFrom, useHardClipping);
                     }
                 }
             }
         }
     }
 
-    /** Returns the number of soft-clipped bases until a non-soft-clipping element is encountered. */
-    private static int lengthOfSoftClipping(Iterator<CigarElement> iterator) {
-        int clipped = 0;
-        while (iterator.hasNext()) {
-            final CigarElement elem = iterator.next();
-            if (elem.getOperator() != CigarOperator.SOFT_CLIP && elem.getOperator() != CigarOperator.HARD_CLIP) {
-                break;
-            }
-            if (elem.getOperator() == CigarOperator.SOFT_CLIP) {
-                clipped = elem.getLength();
+    private static int getReadPositionAtReferencePositionIgnoreSoftClips(final SAMRecord rec, final int pos) {
+        final int readPosition;
+        final Cigar oldCigar = rec.getCigar();
+        final int oldStart = rec.getAlignmentStart();
+        final Cigar newCigar = new Cigar();
+        final List<CigarElement> cigarElements = new ArrayList<>(oldCigar.getCigarElements());
+        int posShift = 0;
+        boolean foundNonClip = false;
+
+        for (final CigarElement cigarElement : cigarElements) {
+            final CigarOperator op = cigarElement.getOperator();
+
+            // Replace SOFT_CLIPs in the oldCigar with MATCH_OR_MISMATCH in newCigar
+            if (op == CigarOperator.SOFT_CLIP) {
+                newCigar.add(new CigarElement(cigarElement.getLength(), CigarOperator.MATCH_OR_MISMATCH));
+                if (!foundNonClip) {
+                    posShift += cigarElement.getLength();
+                }
+            } else {
+                if (!op.isClipping()) {
+                    foundNonClip = true;
+                }
+                newCigar.add(new CigarElement(cigarElement.getLength(), op));
             }
         }
 
-        return clipped;
+        // Temporarily use the newCigar that has SOFT_CLIPs replaced with MATCH_OR_MISMATCH to get read position at reference, but ignore existence of soft-clips
+        rec.setCigar(newCigar);
+        readPosition = SAMRecord.getReadPositionAtReferencePosition(rec, pos, false);
+        rec.setCigar(oldCigar);
+        return readPosition + posShift;
```

Comment:
`// instead of setting back the position of the read by posShift, which could create a negative start position, we add posShift the final position in the read.`